### ISSUE TITLE
sepEndBy'

### ISSUE THREAD

**Issue by jamesdbrock** [jamesdbrock/replace-megaparsec#10]
Implement

    sepEndBy' :: MonadPlus m => m a -> m sep -> m ([a], Maybe sep)

http://hackage.haskell.org/package/parser-combinators-1.0.2/docs/Control-Monad-Combinators.html#v:sepEndBy

or

    skipManyTill' :: MonadPlus m => m a -> m end -> m ([a], Maybe end)

**Comment by jamesdbrock** [comment]
or 
```haskell
sepEndBy' :: m a -> m (Either String a)
sepEndBy'  sep =  sep  <|> (fmap mconcat $ manyTill anyChar sep)
```
inefficient because sep has to match twice for each pattern, but still maybe something useful here?

**Comment by jamesdbrock** [comment]
how about `skipManyTill`?

https://www.stackage.org/haddock/nightly-2019-09-19/parser-combinators/Control-Monad-Combinators.html#v:skipManyTill

```haskell
sepEndBy' :: m a -> m (String, a)
sepEndBy' sep = do
  getOffset <- off1
  skipManyTill $ do
    off2 <- getOffset
    x <- sep
    return (stringSectionBetween off1 off2, x)
    // and what about when we come to the end of the stream?
```

**Comment by jamesdbrock** [comment]
https://www.stackage.org/haddock/nightly-2019-09-19/megaparsec-7.0.5/src/Text.Megaparsec.Internal.html#pTakeWhileP

**Comment by jamesdbrock** [comment]
See also @chshersh 's issue:

https://github.com/mrkkrp/megaparsec/issues/367

**Comment by jamesdbrock** [comment]
So here's the thing. Parsers such as `takeWhileP` are based on 
[`span`](http://hackage.haskell.org/package/text-1.2.4.0/docs/Data-Text.html#v:span)
, which has type 

```haskell
span :: (Char -> Bool) -> Text -> (Text, Text) 
```

The problem is the predicate only takes one char.

There is also [`breakOn`](http://hackage.haskell.org/package/text-1.2.4.0/docs/Data-Text.html#v:breakOn) which is predicated on a constant string.

What we need to write a `sepCap` which is fast in the sparse case is a function which doesn't seems to exist in the world of `Text` or `ByteString`, which is like `breakOn`, but predicated on the beginning of the second string in the break.

```haskell
breakPred :: (Text -> Bool) -> Text -> (Text, Text)
```

Actually, what we want is not a predicate, because we need information about how much of the string to break on. So we need this function.

```haskell
breakSatisfy :: (Text -> Maybe Int) -> Text -> (Text, Text, Text)
```

Let's call `(Text -> Maybe Int)` the "satisfaction" function. `breakSatisfy` will call the satisfaction function on each tail of the input string until the satisfaction function returns a span. Then `breakSatisfy` will return the substring up until satisfaction began, the satisfying substring, and the remaining unsatisfying substring.

**Comment by jamesdbrock** [comment]
```haskell
untilCap :: Parser a -> Parser (Text, a)
```

Parser fails when no match. Prefix or suffix may be zero length. Zero-width match allowed?

```haskell
streamSplit :: Parser a -> s -> Maybe (Text, a, Text)
```

Nothing when `untilCap` fails.

**Comment by jamesdbrock** [comment]
@mrkkrp mostly solved this problem this summer in v1.2.0, see [`manyTill_`](http://hackage.haskell.org/package/parser-combinators-1.2.0/docs/Control-Monad-Combinators.html#v:manyTill_) and `someTill_`.

**Comment by jamesdbrock** [comment]
Added [`anyTill`](https://hackage.haskell.org/package/replace-megaparsec-1.4.1.0/docs/Replace-Megaparsec.html#v:anyTill) for `untilCap`

Added [`splitCap`](https://hackage.haskell.org/package/replace-megaparsec-1.4.1.0/docs/Replace-Megaparsec.html#v:splitCap) for `streamSplit`
